File: ts/examples/basic-usage.ts
/**
 * Basic Lore usage — publish, query, and format lessons.
 *
 * Note: This example uses a fake embedding function for demonstration.
 * Replace with a real embedding model (OpenAI, Cohere, etc.) in production.
 */

import { Lore } from 'lore-sdk';
import { createHash } from 'crypto';

// Simple deterministic embedding for demonstration
function fakeEmbedding(text: string): number[] {
  const hash = createHash('sha256').update(text).digest();
  return Array.from({ length: 384 }, (_, i) => (hash[i % 32] - 128) / 128);
}

async function main() {
  const lore = new Lore({
    dbPath: '/tmp/lore_ts_example.db',
    embeddingFn: fakeEmbedding,
  });

  // Publish lessons
  await lore.publish({
    problem: 'Stripe API returns 429 after 100 requests/min',
    resolution: 'Add exponential backoff starting at 1s, cap at 32s',
    tags: ['stripe', 'rate-limit'],
    confidence: 0.9,
  });

  await lore.publish({
    problem: 'OpenAI API times out on large prompts',
    resolution: 'Split into chunks of 50K tokens',
    tags: ['openai', 'timeout'],
    confidence: 0.8,
  });

  // Query
  const results = await lore.query('how to handle API rate limits');
  console.log(`Found ${results.length} results:\n`);

  for (const r of results) {
    console.log(`  [${r.score.toFixed(3)}] ${r.lesson.problem}`);
    console.log(`           → ${r.lesson.resolution}\n`);
  }

  // Format for prompt
  const prompt = lore.asPrompt(results);
  console.log('--- Prompt section ---');
  console.log(prompt);

  await lore.close();
}

main().catch(console.error);

File: examples/basic_usage.py
"""Basic Lore usage — publish, query, and format lessons."""

from lore import Lore

# Initialize with defaults (SQLite at ~/.lore/default.db)
# Using a temp path for this example; omit db_path for the default location.
lore = Lore(db_path="/tmp/lore_example.db")

# Publish some lessons
lore.publish(
    problem="Stripe API returns 429 after 100 requests/min",
    resolution="Add exponential backoff starting at 1s, cap at 32s",
    tags=["stripe", "rate-limit", "api"],
    confidence=0.9,
)

lore.publish(
    problem="OpenAI API times out on large prompts (>100K tokens)",
    resolution="Split into chunks of 50K tokens, process sequentially",
    tags=["openai", "timeout", "chunking"],
    confidence=0.8,
)

lore.publish(
    problem="PostgreSQL connection pool exhausted under load",
    resolution="Increase pool size to 20, add connection timeout of 30s",
    tags=["postgres", "connection-pool"],
    confidence=0.7,
)

# Query for relevant lessons
results = lore.query("how to handle API rate limits")
print(f"Found {len(results)} results:\n")

for r in results:
    print(f"  [{r.score:.3f}] {r.lesson.problem}")
    print(f"           → {r.lesson.resolution}\n")

# Format for system prompt injection
prompt = lore.as_prompt(results)
print("--- Prompt section ---")
print(prompt)

# Clean up
lore.close()
